src/postgraphql/graphiql/src/components/PostGraphiQL.js

```diff
@@ -0,0 +1,195 @@
+   * update the navigation stack with this function.
+   *
+   * I’m sorry Lee Byron.
+   */
+  // TODO: Submit a PR which adds this as a non-hack.
+  _updateGraphiQLDocExplorerNavStack (nextSchema) {
+    // Get the documentation explorer component from GraphiQL. Unfortunately
+    // for them this looks like public API. Muwahahahaha.
+    const { docExplorerComponent } = this.graphiql
+    const { navStack } = docExplorerComponent.state
+
+    // If one type/field isn’t find this will be set to false and the
+    // `navStack` will just reset itself.
+    let allOk = true
+
+    // Ok, so if you look at GraphiQL source code, the `navStack` is made up of
+    // objects that are either types or fields. Let’s use that to search in
+    // our new schema for matching (updated) types and fields.
+    const nextNavStack = navStack.map((typeOrField, i) => {
+      // If we are not ok, abort!
+      if (!allOk)
+        return null
+
+      // If this is a type, let’s do some shenanigans...
+      if (isType(typeOrField)) {
+        // Let’s see if we can get a type with the same name.
+        const nextType = nextSchema.getType(typeOrField.name)
+
+        // If there is no type with this name (it was removed), we are not ok
+        // so set `allOk` to false and return undefined.
+        if (!nextType) {
+          allOk = false
+          return null
+        }
+
+        // If there is a type with the same name, let’s return it! This is the
+        // new type with all our new information.
+        return nextType
+      }
+      // If you thought this function was already pretty bad, it’s about to get
+      // worse. We want to update the information for an object field.
+      else {
+        // Ok, so since this is an object field, we will assume that the last
+        // element in our stack was an object type.
+        const nextLastType = nextSchema.getType(navStack[i - 1] ? navStack[i - 1].name : null)
+
+        // If there is no type for the last type in the nav stack’s name.
+        // Panic!
+        if (!nextLastType) {
+          allOk = false
+          return null
+        }
+
+        // If the last type is not an object type. Panic!
+        if (!(nextLastType instanceof GraphQLObjectType)) {
+          allOk = false
+          return null
+        }
+
+        // Next we will see if the new field exists in the last object type.
+        const nextField = nextLastType.getFields()[typeOrField.name]
+
+        // If not, Panic!
+        if (!nextField) {
+          allOk = false
+          return null
+        }
+
+        // Otherwise we hope very much that it is correct.
+        return nextField
+      }
+    })
+
+    // This is very hacky but works. React is cool.
+    this.graphiql.docExplorerComponent.setState({
+      // If we are not ok, just reset the `navStack` with an empty array.
+      // Otherwise use our new stack.
+      navStack: allOk ? nextNavStack : [],
+    })
```

Comment:
The versions of both [`graphql` and `graphiql` are locked](https://github.com/calebmer/postgraphql/blob/8f93257769258d39dc294072d4d1d84a9cb02094/src/postgraphql/graphiql/package.json#L10-L11).

The reason we create a new `navStack` is so that the documentation viewer doesn’t reset or have stale data. When we update our schema, vanilla GraphiQL does not update the `navStack` to reflect the changes as the `navStack` is an array of `GraphQLFieldConfig | GraphQLObjectType` from `graphql-js`. The old field configs and object type’s are not replaced with the new field configs and object type’s when the schema prop changes. An example of this is you are on the `Person` type page in the GraphiQL documentation viewer and the schema changes so `Person` has a new description. Without this hack the description in GraphiQL would not change. Same for adding/removing fields.

This is why we manually replace the `navStack`, we can maintain the user’s navigation state while still showing the new data from the schema. Throwing away the GraphiQL instance would lose user navigation state and potentially even their query/variables/response.

I definitely agree though that we should submit a PR to GraphiQL incorporating these changes. I was on the clock to get this done, thus why it is in PostGraphQL :wink:. I’m fine with merging it in as-is with the recognition that we will need to PR GraphiQL next time we touch this logic.